### PR TITLE
Added a docker-compose for hotrod e2e test

### PR DESCRIPTION
**Pull request by hellspawn679**
## Which problem is this PR solving?
- https://github.com/jaegertracing/jaeger/issues/5735

## Description of the changes
-  added docker-compose file to test hotrod so that is can connect to jaeger

## How was this change tested?
- 

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: docker-compose/hotrod/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+version: '3.7'
```
Comment (yurishkuro) [inline docker-compose/hotrod/docker-compose.yml:1]: We already have this in examples/hotrod/docker-compose.yml

**Comment by yurishkuro** [comment]
I suggest you explain your plan, doesn't seem it's going in the right direction. We already have all the pieces you included in this PR

**Comment by hellspawn679** [comment]
@yurishkuro  my plan is to build a local hotrod image(localhost:5000/jaegertracing/example-hotrod:${GITHUB_SHA:-latest}) and test it using docker-compose(that way we won't have any error for port 4318) then we will hit the frontend end point and get the hotrod logs and get the tracesID from there

File: docker-compose/hotrod/docker-compose.yml (hunk 1 above)
Comment (hellspawn679) [reply docker-compose/hotrod/docker-compose.yml:1]: my image tag is different that is why I create a different file 
`localhost:5000/jaegertracing/example-hotrod:${GITHUB_SHA:-latest}`

Comment (yurishkuro) [reply docker-compose/hotrod/docker-compose.yml:1]: this is not the reason to duplicate the docker compose file. You should instead parameterize the main one.

File: scripts/hotrod_dummy_test.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
```
Comment (yurishkuro) [inline scripts/hotrod_dummy_test.sh:1]: we already have a script that does basic hotrod test: `scripts/hotrod-integration-test.sh`. The objective is to make the checks stronger, not to duplicate.

File: examples/hotrod/docker-compose.yml [hunk 3]
```diff
@@ -12,13 +9,13 @@ services:
       - "4318:4318"
     environment:
       - LOG_LEVEL=debug
+      - COLLECTOR_OTLP_GRPC_HOST_PORT=0.0.0.0:4317
+      - COLLECTOR_OTLP_HTTP_HOST_PORT=0.0.0.0:4318
```
Comment (yurishkuro) [inline examples/hotrod/docker-compose.yml:13]: this is no longer necessary, I merged a fix to `main`

File: examples/hotrod/docker-compose.yml [hunk 4]
```diff
@@ -12,13 +9,13 @@ services:
...
     networks:
       - jaeger-example
+
   hotrod:
-    image: jaegertracing/example-hotrod:${JAEGER_VERSION:-latest}
-    # To run the latest trunk build, find the tag at Docker Hub and use the line below
-    # https://hub.docker.com/r/jaegertracing/example-hotrod-snapshot/tags
-    #image: jaegertracing/example-hotrod-snapshot:0ab8f2fcb12ff0d10830c1ee3bb52b745522db6c
+    image: ${HOTROD_IMAGE:-jaegertracing/example-hotrod}:${GITHUB_SHA:-${JAEGER_VERSION:-latest}}
```
Comment (yurishkuro) [inline examples/hotrod/docker-compose.yml:18]: GITHUB_SHA has nothing to do with this script. JAEGER_VERSION variable is sufficient.

Comment (yurishkuro) [reply examples/hotrod/docker-compose.yml:18]: > HOTROD_IMAGE:-jaegertracing/example-hotrod

Are you actually planning to override the name, or just the registry? I think it's only the registry, so specialize the var to that

```
image: ${REGISTRY:-}jaegertracing/example-hotrod:${JAEGER_VERSION:-latest}
```

**Review by yurishkuro — COMMENTED** [review]
you have UI submodule affected in this PR, please remove.

File: examples/hotrod/docker-compose.yml (hunk 3 above)
Comment (yurishkuro) [reply examples/hotrod/docker-compose.yml:15]: if you run it with binaries built from `main` branch it should work.

File: examples/hotrod/docker-compose.yml [hunk 5]
```diff
@@ -14,11 +13,9 @@ services:
       - LOG_LEVEL=debug
     networks:
       - jaeger-example
+
   hotrod:
-    image: jaegertracing/example-hotrod:${JAEGER_VERSION:-latest}
-    # To run the latest trunk build, find the tag at Docker Hub and use the line below
-    # https://hub.docker.com/r/jaegertracing/example-hotrod-snapshot/tags
-    #image: jaegertracing/example-hotrod-snapshot:0ab8f2fcb12ff0d10830c1ee3bb52b745522db6c
```
Comment (yurishkuro) [inline examples/hotrod/docker-compose.yml:21]: why remove the commend? Your change is unrelated to it.

File: scripts/hotrod-integration-test.sh [hunk 6]
```diff
@@ -28,6 +27,39 @@ if [[ $body != *"Rides On Demand"* ]]; then
   echo "String \"Rides On Demand\" is not present on the index page"
   exit 1
 fi
-docker rm -f "$CID"
 
-bash scripts/build-upload-a-docker-image.sh -c example-hotrod -d examples/hotrod -p "${platforms}"
+curl -X POST "http://localhost:8080/dispatch?customer=123"
+# Extract the logs from the docker compose service
+logs=$(docker compose -f ./examples/hotrod/docker-compose.yml logs hotrod)
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:33]: since you're repeating this path, store it in a var once

File: scripts/hotrod-integration-test.sh [hunk 7]
```diff
@@ -7,16 +7,15 @@ make build-examples GOOS=linux GOARCH=s390x
 make build-examples GOOS=linux GOARCH=ppc64le
 make build-examples GOOS=linux GOARCH=arm64
 
-REPO=jaegertracing/example-hotrod
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:10]: why remove it?

File: scripts/hotrod-integration-test.sh [hunk 8]
```diff
@@ -7,16 +7,15 @@ make build-examples GOOS=linux GOARCH=s390x
...
 platforms="linux/amd64,linux/s390x,linux/ppc64le,linux/arm64"
 make prepare-docker-buildx
 
 # build image locally (-l) for integration test
 bash scripts/build-upload-a-docker-image.sh -l -c example-hotrod -d examples/hotrod -p "${platforms}"
 
-# pass --name example-hotrod so that we can do `docker logs example-hotrod` later
-export CID
-CID=$(docker run -d --name example-hotrod -p 8080:8080 "localhost:5000/${REPO}:${GITHUB_SHA}")
+export HOTROD_IMAGE="localhost:5000/jaegertracing/example-hotrod"
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:17]: `HOTROD_IMAGE` is not used in the compose file

File: scripts/hotrod-integration-test.sh [hunk 9]
```diff
@@ -28,6 +27,39 @@ if [[ $body != *"Rides On Demand"* ]]; then
...
 fi
-docker rm -f "$CID"
 
-bash scripts/build-upload-a-docker-image.sh -c example-hotrod -d examples/hotrod -p "${platforms}"
+curl -X POST "http://localhost:8080/dispatch?customer=123"
+# Extract the logs from the docker compose service
+logs=$(docker compose -f ./examples/hotrod/docker-compose.yml logs hotrod)
+
+# Extract the trace_id from the logs
+TRACE_ID=$(echo "$logs" | grep -oP '(?<="trace_id": ")[^"]+' | tail -n 1)
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:36]: the trace ID should be read from curl request response headers, not from logs

File: scripts/hotrod-integration-test.sh [hunk 10]
```diff
@@ -28,6 +27,39 @@ if [[ $body != *"Rides On Demand"* ]]; then
   echo "String \"Rides On Demand\" is not present on the index page"
   exit 1
 fi
-docker rm -f "$CID"
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:31]: The test needs to shut down containers once it's done, you are leaving them running.

File: scripts/hotrod-integration-test.sh (hunk 7 above)
Comment (hellspawn679) [reply scripts/hotrod-integration-test.sh:10]: that wasn't being used any where

File: examples/hotrod/docker-compose.yml (hunk 4 above)
Comment (hellspawn679) [reply examples/hotrod/docker-compose.yml]: > GITHUB_SHA has nothing to do with this script. JAEGER_VERSION variable is sufficient.

but the local image build by `scripts/build-upload-a-docker-image.sh` gives it a tag like this so that is why I need GITHUB_SHA
`"localhost:5000/${namespace}/${component_name}:${GITHUB_SHA}"`

Comment (yurishkuro) [reply examples/hotrod/docker-compose.yml]: It doesn't matter if another script needs it, that should be taken care of as an integration point. This docker-compose itself has no need for GITHUB_SHA, it only needs JAEGER_VERSION

Comment (hellspawn679) [reply examples/hotrod/docker-compose.yml]: but the local image that is created for testing  has name 
localhost:5000/jaegertracing/example-hotrod:${GITHUB_SHA}
so I want to test using this image I need to use :${GITHUB_SHA} in docker compose otherwise it will pull the latest image instead of using the local build binary

Comment (yurishkuro) [reply examples/hotrod/docker-compose.yml]: So call the compose file with `JAEGER_VERSION=${GITHUB_SHA}` env var.

Comment (hellspawn679) [reply examples/hotrod/docker-compose.yml]: but then it will also use it for ` jaegertracing/all-in-one:${JAEGER_VERSION:-latest}` which will give error

File: examples/hotrod/docker-compose.yml [hunk 11]
```diff
@@ -1,11 +1,10 @@
 version: '3.7'
-
 # To run a specific version of Jaeger, use environment variable, e.g.:
-#     JAEGER_VERSION=1.52 docker compose up
+#     JAEGER_VERSION=1.52 docker compose up
```
Comment (yurishkuro) [inline examples/hotrod/docker-compose.yml:3]:
```suggestion
#     JAEGER_VERSION=1.52 docker compose up
```

File: scripts/hotrod-integration-test.sh [hunk 12]
```diff
@@ -2,32 +2,76 @@
...
+EXPECTED_SPANS=10
+MAX_RETRIES=30
+SLEEP_INTERVAL=10
+
+# Function to poll Jaeger for the trace
+poll_jaeger() {
+  local trace_id=$1
+  local url="${JAEGER_QUERY_URL}/api/traces/${trace_id}"
+
+  curl -s "${url}" | jq '.data[0].spans | length'
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:60]: what happens if trace is not found?

File: scripts/hotrod-integration-test.sh [hunk 13]
```diff
@@ -2,32 +2,76 @@
...
 body=$(curl localhost:8080)
 if [[ $body != *"Rides On Demand"* ]]; then
   echo "String \"Rides On Demand\" is not present on the index page"
   exit 1
 fi
-docker rm -f "$CID"
 
-bash scripts/build-upload-a-docker-image.sh -c example-hotrod -d examples/hotrod -p "${platforms}"
+response=$(curl -i -X POST "http://localhost:8080/dispatch?customer=123")
+TRACE_ID=$(echo "$response" | grep -Fi "Traceresponse" | awk '{print $2}' | cut -d '-' -f 2)
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:48]: check that it's not empty?

File: scripts/hotrod-integration-test.sh [hunk 14]
```diff
@@ -2,32 +2,76 @@
...
+
+JAEGER_QUERY_URL="http://localhost:16686"
+EXPECTED_SPANS=10
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:51]:
```suggestion
EXPECTED_SPANS=35
```

File: scripts/hotrod-integration-test.sh (hunk 12 above)
Comment (yurishkuro) [reply scripts/hotrod-integration-test.sh:60]: in other words, wouldn't you get a pipe failure, e.g. if jq cannot parse? What will happen to the overall script in this case?

**Review by yurishkuro — COMMENTED** [review]
lgtm, a couple open questions

File: scripts/hotrod-integration-test.sh (hunk 12 above)
Comment (hellspawn679) [reply scripts/hotrod-integration-test.sh:60]: the script will show this error and shut down 
image

Comment (yurishkuro) [reply scripts/hotrod-integration-test.sh:67]: that's a problem - your run this function inside a loop waiting for a condition to be met. Failing to meet the condition in one of the iterations is not a fatal error, it should not abort the whole script.

File: scripts/hotrod-integration-test.sh [hunk 15]
```diff
@@ -2,32 +2,83 @@
...
 body=$(curl localhost:8080)
 if [[ $body != *"Rides On Demand"* ]]; then
   echo "String \"Rides On Demand\" is not present on the index page"
   exit 1
 fi
-docker rm -f "$CID"
 
-bash scripts/build-upload-a-docker-image.sh -c example-hotrod -d examples/hotrod -p "${platforms}"
+response=$(curl -i -X POST "http://localhost:8080/dispatch?customer=123")
+TRACE_ID=$(echo "$response" | grep -Fi "Traceresponse" | awk '{print $2}' | cut -d '-' -f 2)
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:48]:
```suggestion
TRACE_ID=$(echo "$response" | grep -Fi "Traceresponse:" | awk '{print $2}' | cut -d '-' -f 2)
```
can you grep with `:`? This would indicate that you're reading a header, not just some content of the response.

File: scripts/hotrod-integration-test.sh [hunk 16]
```diff
@@ -2,32 +2,83 @@
...
+EXPECTED_SPANS=35
+MAX_RETRIES=30
+SLEEP_INTERVAL=10
+
+# Function to poll Jaeger for the trace
+poll_jaeger() {
+  local trace_id=$1
+  local url="${JAEGER_QUERY_URL}/api/traces/${trace_id}"
+
+  curl -s "${url}" | jq '.data[0].spans | length' ||  0
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:67]: `||` is supposed to be followed by a command. Which command is `0`?

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
Thanks!

File: scripts/hotrod-integration-test.sh [hunk 17]
```diff
@@ -2,32 +2,83 @@
...
 done
+
 body=$(curl localhost:8080)
 if [[ $body != *"Rides On Demand"* ]]; then
   echo "String \"Rides On Demand\" is not present on the index page"
   exit 1
 fi
-docker rm -f "$CID"
 
-bash scripts/build-upload-a-docker-image.sh -c example-hotrod -d examples/hotrod -p "${platforms}"
```
Comment (yurishkuro) [inline scripts/hotrod-integration-test.sh:33]: this line should not have been deleted. It needs to run as the last step of the script.
